tests/unit/lms/services/course_test.py

```diff
@@ -50,6 +51,73 @@ def test_get_by_context_id_with_no_match_and_raise_on_missing(self, svc):
         with pytest.raises(NoResultFound):
             svc.get_by_context_id("NO MATCH", raise_on_missing=True)
 
+    def test_get_from_launch_when_existing(
+        self, svc, get_by_context_id, upsert_course, product
+    ):
+        course = get_by_context_id.return_value = factories.Course(
+            extra={"existing": "extra"}
+        )
+
+        course = svc.get_from_launch(
+            product,
+            lti_params={
+                "context_id": sentinel.context_id,
+                "context_title": sentinel.context_title,
+            },
+        )
+
+        get_by_context_id.assert_called_once_with(
+            sentinel.context_id,
+        )
+        upsert_course.assert_called_once_with(
+            context_id=sentinel.context_id,
+            name=sentinel.context_title,
+            extra={"existing": "extra"},
+        )
+        assert course == upsert_course.return_value
+
+    def test_get_from_launch_when_new(
+        self, svc, get_by_context_id, upsert_course, product
+    ):
+        get_by_context_id.return_value = None
+
+        course = svc.get_from_launch(
+            product,
+            lti_params={
+                "context_id": sentinel.context_id,
+                "context_title": sentinel.context_title,
+            },
+        )
+
+        get_by_context_id.assert_called_once_with(sentinel.context_id)
+        upsert_course.assert_called_once_with(
+            context_id=sentinel.context_id, name=sentinel.context_title, extra={}
+        )
+        assert course == upsert_course.return_value
+
+    def test_get_from_launch_when_new_and_canvas(
+        self, svc, upsert_course, get_by_context_id, product
+    ):
+        get_by_context_id.return_value = None
+        product.family = Product.Family.CANVAS
+
+        course = svc.get_from_launch(
+            product,
+            lti_params={
+                "context_id": sentinel.context_id,
+                "context_title": sentinel.context_title,
+                "custom_canvas_course_id": sentinel.canvas_id,
+            },
+        )
+
+        get_by_context_id.assert_called_once_with(sentinel.context_id)
+        upsert_course.assert_called_once_with(
+            context_id=sentinel.context_id,
+            name=sentinel.context_title,
+            extra={"canvas": {"custom_canvas_course_id": sentinel.canvas_id}},
+        )
+        assert course == upsert_course.return_value
```

Comment:
These two look potentially parameterizeable with `family,extra_lti_params,extra` or something. Not sure if it would be longer though.